qos/evm/endpoint.go

```diff
@@ -1,100 +1,148 @@
+	for _, check := range e.checks {
+		if err := check.isValid(serviceState); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
 // ApplyObservation updates the data stored regarding the endpoint using the supplied observation.
-// It Returns true if the observation was not unrecognized, i.e. mutated the endpoint.
+// It returns true if the observation was not unrecognized, i.e. mutated the endpoint.
 // TODO_TECHDEBT(@adshmh): add a method to distinguish the following two scenarios:
 //   - an endpoint that returned in invalid response.
 //   - an endpoint with no/incomplete observations.
 func (e *endpoint) ApplyObservation(obs *qosobservations.EVMEndpointObservation) bool {
+	// If emptyResponse is not nil, the observation is for an empty response check.
 	if obs.GetEmptyResponse() != nil {
-		e.hasReturnedEmptyResponse = true
+		e.applyEmptyResponseObservation()
 		return true
 	}
 
+	// If chainIDResponse is not nil, the observation is for a chainID check.
 	if chainIDResponse := obs.GetChainIdResponse(); chainIDResponse != nil {
-		observedChainID := chainIDResponse.GetChainIdResponse()
-		e.chainIDResponse = &observedChainID
+		e.applyChainIDObservation(chainIDResponse)
 		return true
 	}
 
+	// If blockNumberResponse is not nil, the obbservation is for a blockNumber check.
 	if blockNumberResponse := obs.GetBlockNumberResponse(); blockNumberResponse != nil {
-		// base 0: use the string's prefix to determine its base.
-		parsedBlockNumber, err := strconv.ParseUint(blockNumberResponse.GetBlockNumberResponse(), 0, 64)
-		// The endpoint returned an invalid response to an `eth_blockNumber` request.
-		// Explicitly set the parsedBlockNumberResponse to a zero value as the ParseUInt does not guarantee returning a 0 on all error cases.
-		if err != nil {
-			zero := uint64(0)
-			e.parsedBlockNumberResponse = &zero
-			return true
-		}
-
-		e.parsedBlockNumberResponse = &parsedBlockNumber
+		e.applyBlockNumberObservation(blockNumberResponse)
 		return true
 	}
 
 	return false
 }
 
+// applyEmptyResponseObservation updates the empty response check if a valid observation is provided.
+func (e *endpoint) applyEmptyResponseObservation() {
+	e.checks[checkNameEmptyResponse].check = &endpointCheckEmptyResponse{
+		hasReturnedEmptyResponse: true, // An empty response is always invalid.
+	}
+}
+
+// applyChainIDObservation updates the chain ID check if a valid observation is provided.
+func (e *endpoint) applyChainIDObservation(chainIDResponse *qosobservations.EVMChainIDResponse) {
+	observedChainID := chainIDResponse.GetChainIdResponse()
+
+	e.checks[checkNameChainID].check = &endpointCheckChainID{
+		chainID:   &observedChainID,
+		expiresAt: time.Now().Add(checkChainIDInterval),
+	}
+}
+
+// applyBlockNumberObservation updates the block number check if a valid observation is provided.
+func (e *endpoint) applyBlockNumberObservation(blockNumberResponse *qosobservations.EVMBlockNumberResponse) {
```

Comment:
I kind of feel like these functions should live in the corresponding `check_XXX.go` file.

I know the receiver is an `endpoint` but it does seem a bit weird to centralize it.

Not a strong opinion here but just a gut feeling.